arangod/Aql/Functions.cpp

```diff
@@ -6547,6 +6547,354 @@ AqlValue Functions::Pi(ExpressionContext*, transaction::Methods* trx,
+    if (index < Functions::bitFunctionsMaxSupportedBits) {
+      return AqlValue(AqlValueHintBool((testee & (uint64_t(1) << index)) != 0));
+    }
+  }
+      
+  static char const* AFN = "BIT_TEST";
+  registerInvalidArgumentWarning(expressionContext, AFN);
+  return AqlValue(AqlValueHintNull());
+}
+
+/// @brief function BIT_SHIFT_LEFT
+AqlValue Functions::BitShiftLeft(ExpressionContext* expressionContext, transaction::Methods*,
+                                 VPackFunctionParameters const& parameters) {
+  auto result = binaryBitFunctionParameters<uint64_t, uint64_t>(parameters);
+  if (result.has_value()) {
+    auto [testee, shift] = result.value();
+    AqlValue const& value = extractFunctionParameterValue(parameters, 2);
+    if (value.isNumber()) {
+      auto width = bitOperationValue<uint64_t>(value.slice());
+      if (width.has_value() && width.value() <= static_cast<uint64_t>(Functions::bitFunctionsMaxSupportedBits)) {
+        // mask the lower bits of the result with up to 32 active bits
+        return AqlValue(AqlValueHintUInt((testee << shift) & ((uint64_t(1) << width.value()) - 1)));
+      }
+    }
+  }
+  
+  static char const* AFN = "BIT_SHIFT_LEFT";
+  registerInvalidArgumentWarning(expressionContext, AFN);
+  return AqlValue(AqlValueHintNull());
+}
+
+/// @brief function BIT_SHIFT_RIGHT
+AqlValue Functions::BitShiftRight(ExpressionContext* expressionContext, transaction::Methods*,
+                                  VPackFunctionParameters const& parameters) {
+  auto result = binaryBitFunctionParameters<uint64_t, uint64_t>(parameters);
+  if (result.has_value()) {
+    auto [testee, shift] = result.value();
+    AqlValue const& value = extractFunctionParameterValue(parameters, 2);
+    if (value.isNumber()) {
+      auto width = bitOperationValue<uint32_t>(value.slice());
+      if (width.has_value() && width.value() <= static_cast<uint64_t>(Functions::bitFunctionsMaxSupportedBits)) {
+        // mask the lower bits of the result with up to 32 active bits
+        return AqlValue(AqlValueHintUInt((testee >> shift) & ((uint64_t(1) << width.value()) - 1)));
+      }
+    }
+  }
+  
+  static char const* AFN = "BIT_SHIFT_RIGHT";
+  registerInvalidArgumentWarning(expressionContext, AFN);
+  return AqlValue(AqlValueHintNull());
+}
+  
+/// @brief function BIT_POPCOUNT
+AqlValue Functions::BitPopcount(ExpressionContext* expressionContext, transaction::Methods* trx,
+                                VPackFunctionParameters const& parameters) {
+  AqlValue const& value = extractFunctionParameterValue(parameters, 0);
+
+  if (value.isNumber()) {
+    VPackSlice v = value.slice();
+    auto result = bitOperationValue<uint64_t>(v);
+    if (result.has_value()) {
+      uint64_t x = result.value();
+      // the following code fragment is from https://en.wikipedia.org/wiki/Hamming_weight#Efficient_implementation
+      // TODO: we could optimize this to use hardware-accelerated popcnt if it exists, or gcc's __builtin_popcntll
```

Comment:
FYI
iresearch has `popcnt` implemented using compiler intrinsics (`utils/math_utils.hpp`), you can use it like
```
irs::math_utils::math_traits<decltype(x)>::pop(x)
```

Reply:
Ok, will do this in devel. Thanks for the info!

Reply:
Done in devel. Leaving the 3.7 as it is (also covered by tests, so results are also correct for 3.7).